core/io/resource_loader.cpp

```diff
@@ -1201,6 +1257,58 @@ bool ResourceLoader::should_create_uid_file(const String &p_path) {
 	return false;
 }
 
+bool ResourceLoader::_is_path_whitelisted(const String &p_path, const Dictionary &p_whitelist) {
+	if (p_whitelist.is_empty()) {
+		return false;
+	}
+
+	// Security: Validate input path doesn't contain null bytes (defense in depth)
+	// Null bytes can be used in path traversal attacks (e.g., "../../etc/passwd%00.png")
+	// Even though Godot's String class handles null bytes, we explicitly reject them for security
+	if (p_path.find_char('\0') >= 0) {
+		return false;
+	}
+
+	// Security: Validate whitelist dictionary structure
+	// Ensure all keys are strings to prevent type confusion attacks
+	Array keys = p_whitelist.keys();
+	for (int i = 0; i < keys.size(); i++) {
+		Variant key = keys[i];
+		// Only allow string keys in whitelist dictionary
+		if (key.get_type() != Variant::STRING) {
+			return false;
+		}
+	}
+
+	// Normalize the input path to prevent path traversal attacks
+	// This ensures paths like "res://textures/../secret/file.png" are properly normalized
+	// before comparison, preventing them from matching "res://textures/"
+	String normalized_path = p_path.simplify_path();
+
+	// Check for exact match first (backward compatible)
+	// Check both original and normalized path for exact matches
+	if (p_whitelist.has(p_path) || p_whitelist.has(normalized_path)) {
```

Comment:
Not sure what backwards compatible means here, as this is the first time whitelist is impl. I would only check for normalized path.